File: XUCore/additions/NSMutableURLRequestAdditions.m
// 
// NSMutableURLRequestAdditions.m
// 
// Created by Charlie Monroe
// 
// Copyright (c) 2010-14 Charlie Monroe Software. All rights reserved.
// 


#import "NSMutableURLRequestAdditions.h"

#import "NSArrayAdditions.h"
#import "NSDataAdditions.h"
#import "NSStringAdditions.h"

@implementation NSMutableURLRequest (Additions)

-(void)addAccept:(NSString*)accept{
	[self addValue:accept forHTTPHeaderField:@"Accept"];
}
-(void)addContentType:(NSString*)contentType{
	[self addValue:contentType forHTTPHeaderField:@"Content-Type"];
}
-(void)addJSONAcceptToHeader{
	[self addAccept:@"application/json"];
}
-(void)addJSONContentToHeader{
	[self addContentType:@"application/json"];
}
-(void)addMultipartFormDataContentToHeader{
	[self addContentType:@"multipart/form-data"];
}
-(void)addWWWFormContentToHeader{
	[self addContentType:@"application/x-www-form-urlencoded"];
}
-(void)addXMLAcceptToHeader{
	[self addAccept:@"application/xml"];
}
-(void)addXMLContentToHeader{
	[self addContentType:@"application/xml"];
}
-(NSString *)referer{
	return [self valueForHTTPHeaderField:@"Referer"];
}
-(void)setFormBody:(NSDictionary *)formBody{
	NSString *bodyString = [[[formBody allKeys] map:^id(NSString *key) {
		return [NSString stringWithFormat:@"%@=%@", key, [[formBody objectForKey:key] stringByEncodingIllegalURLCharacters]];
	}] componentsJoinedByString:@"&"];
	[self setHTTPBody:[bodyString dataUsingEncoding:NSUTF8StringEncoding]];
}
-(void)setJSONBody:(id)obj{
	[self setHTTPBody:[NSJSONSerialization dataWithJSONObject:obj options:0 error:NULL]];
}
-(void)setReferer:(NSString *)referer{
	[self setValue:referer forHTTPHeaderField:@"Referer"];
}
-(void)setUserAgent:(NSString *)useragent{
	[self setValue:useragent forHTTPHeaderField:@"User-Agent"];
}
-(void)setUsername:(NSString*)name andPassword:(NSString*)password{
	NSString *b64 = [[[NSString stringWithFormat:@"%@:%@", name, password] dataUsingEncoding:NSUTF8StringEncoding] base64EncodedStringWithOptions:0];
	[self addValue:[NSString stringWithFormat:@"Basic %@", b64] forHTTPHeaderField:@"Authorization"];
}
-(NSString *)userAgent{
	return [self valueForHTTPHeaderField:@"User-Agent"];
}

@end

File: XUCore/additions/NSMutableURLRequestAdditions.swift
//
//  NSMutableURLRequestAdditions.swift
//  XUCore
//
//  Created by Charlie Monroe on 11/10/15.
//  Copyright © 2015 Charlie Monroe Software. All rights reserved.
//

import Foundation

/// Value for the accept/content header field.
@available(*, deprecated, renamed="NSURLRequest.ContentType.JSON")
public let XUMutableURLRequestJSONHeaderFieldValue = "application/json;charset=UTF-8"

/// Value for the accept/content header field.
@available(*, deprecated, renamed="NSURLRequest.ContentType.XML")
public let XUMutableURLRequestXMLHeaderFieldValue = "application/xml;charset=UTF-8"

/// Value for the accept/content header field.
@available(*, deprecated, renamed="NSURLRequest.ContentType.WWWForm")
public let XUMutableURLRequestWWWFormHeaderFieldValue = "application/x-www-form-urlencoded;charset=UTF-8"

public extension NSURLRequest {
	
	public struct ContentType {
		
		/// Content-Type: application/json
		public static let JSON = "application/json;charset=UTF-8"
		
		/// Content-Type: application/xml
		public static let XML = "application/xml;charset=UTF-8"
		
		/// Content-Type: application/x-www-form-urlencoded
		public static let WWWForm = "application/x-www-form-urlencoded;charset=UTF-8"
		
	}
	
}


public extension NSMutableURLRequest {
	
	public func addAccept(accept: String) {
		self.addValue(accept, forHTTPHeaderField: "Accept")
	}
	public func addContentType(contentType: String) {
		self.addValue(contentType, forHTTPHeaderField: "Content-Type")
	}
	public func addJSONAcceptToHeader() {
		self.addAccept(NSURLRequest.ContentType.JSON)
	}
	public func addJSONContentToHeader() {
		self.addContentType(NSURLRequest.ContentType.JSON)
	}
	public func addMultipartFormDataContentToHeader() {
		self.addContentType("multipart/form-data")
	}
	public func addWWWFormContentToHeader() {
		self.addContentType(NSURLRequest.ContentType.WWWForm)
	}
	public func addXMLAcceptToHeader() {
		self.addAccept(NSURLRequest.ContentType.XML)
	}
	public func addXMLContentToHeader() {
		self.addContentType(NSURLRequest.ContentType.XML)
	}
	
	public var acceptType: String? {
		get {
			return self.valueForHTTPHeaderField("Accept")
		}
		set {
			self.setValue(newValue, forHTTPHeaderField: "Accept")
		}
	}
	
	public var contentType: String? {
		get {
			return self.valueForHTTPHeaderField("Content-Type")
		}
		set {
			self.setValue(newValue, forHTTPHeaderField: "Content-Type")
		}
	}
	
	public var referer: String? {
		get {
			return self.valueForHTTPHeaderField("Referer")
		}
		set {
			self.setValue(newValue, forHTTPHeaderField: "Referer")
		}
	}
	
	public func setFormBody(formBody: [String : String]) {
		let bodyString = formBody.URLQueryString
		self.HTTPBody = bodyString.dataUsingEncoding(NSUTF8StringEncoding)
	}
	public func setJSONBody(obj: AnyObject) {
		self.HTTPBody = try? NSJSONSerialization.dataWithJSONObject(obj, options: NSJSONWritingOptions())
	}
	public func setUsername(name: String, andPassword password: String) {
		guard let b64 = "\(name):\(password)".dataUsingEncoding(NSUTF8StringEncoding)?.base64EncodedStringWithOptions(NSDataBase64EncodingOptions()) else {
			XULog("Failed to set name and password - cannot create a base64-encoded string!")
			return
		}
		self.addValue("Basic \(b64)", forHTTPHeaderField: "Authorization")
	}
	
	/// You can subscript the URL request and get/set HTTP header fields.
	public subscript(field: String) -> String? {
		get {
			return self.valueForHTTPHeaderField(field)
		}
		set {
			self.setValue(newValue, forHTTPHeaderField: field)
		}
	}
	
	public var userAgent: String? {
		get {
			return self.valueForHTTPHeaderField("User-Agent")
		}
		set {
			self.setValue(newValue, forHTTPHeaderField: "User-Agent")
		}
	}
}

